psa-crypto/src/types/key.rs

```diff
@@ -305,6 +305,37 @@ impl Attributes {
         get_attributes_res?;
         Ok(attributes?)
     }
+
+    /// Sufficient size for a buffer to export the key, if supported
+    #[cfg(feature = "with-mbed-crypto")]
+    pub fn export_key_output_size(self) -> Result<usize> {
+        match self.key_type {
+            Type::RsaKeyPair
+            | Type::RsaPublicKey
+            | Type::EccKeyPair { .. }
+            | Type::EccPublicKey { .. } => Ok(psa_crypto_sys::PSA_EXPORT_KEY_OUTPUT_SIZE(
+                self.key_type.try_into()?,
+                self.bits,
+            )),
+            _ => Err(Error::NotSupported),
+        }
+    }
+
+    /// Sufficient buffer size for a signature using the given key, if the key is supported
+    #[cfg(feature = "with-mbed-crypto")]
+    pub fn sign_output_size(self) -> Result<usize> {
```

Comment:
I think the `alg` parameter of `PSA_SIGN_OUTPUT_SIZE` macro should not be the same as the `permitted_algorithms` field of the attributes but should come from as another `alg` parameter of this function.
My understanding is that the `permitted_algorithms` field only specify the authorised algorithms (there can be multiple) that these attributes allow as opposed as the specific algorithm to use for a particular operation.
Even if the results will be the same, I think it would be better this way, to add an `alg` parameter (of type `AsymmetricSignature`) to this function!

Reply:
I was wondering about that, there seemed to be no specific algorithm for a given Attributes struct. Looking at the way it used to work, it did use `permitted_algorithm`,  just like I've done (it converted `Attributes` into `psa_key_attributes_t` which does `attributes.policy.permitted_algorithms.try_into()?` to get the algorithm). 

Can we not fish out the `AsymmetricSignature` from self and if it doesn't have one, return the error as it does currently?
Something like:
```rust
if let Algorithm::AsymmetricSignature(sign_alg) = self.policy.permitted_algorithms {
    Ok(psa_crypto_sys::PSA_SIGN_OUTPUT_SIZE(
        self.key_type.try_into()?,
        self.bits,
        sign_alg.into(),
    ))
}
else {
    Err(Error::InvalidArgument)
}
```

Reply:
I think that before we were doing things wrong, mixing and using the policy algorithm as the algorithm used for any crypto operation. We [changed that and added an alg field](https://github.com/parallaxsecond/parsec-operations/commit/d301c62d88c10db7501f2eed8bcba17ab0c24202) to specifically split those two semantic units which are the policy algorithms (the algorithms permitted to be used with a key, they can be multiple because you can have `SignHash::Any` as an hash algorithm) and the specific algorithm used for a crypto operation. This last one, [can not be a policy algorithm](https://armmbed.github.io/mbed-crypto/html/api/ops/sign.html#c.PSA_ALG_ANY_HASH).

Following the same logic, I think it would be better splitting things here as well, as in not using the policy algorithm as the one to use to check the output size.

Reply:
Ah ok. So where should the algorithm come from? Should there be another `Algorithm` attribute in the struct or is it held separately?

Reply:
I think it should be part of the method's signature!

```rust
pub fn sign_output_size(self, alg: AsymmetricSignature) -> Result<usize>
```

Reply:
Yeah, I meant where does the caller get the value from :laughing:. I've got it though, `op.alg`.